### PR TITLE
 feat: Add update method for user notification

### PR DESCRIPTION
**Pull request by lynnagara**
Depends on https://github.com/getsentry/sentry/pull/6827

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/api/endpoints/user_notification_details.py [hunk 1]
```diff
@@ -1,14 +1,46 @@
...
+        'key': 'subscribe_by_default',
+        'default': UserOptionValue.participating_only,  # '1'
+        'type': bool,
+    },
+    'workflowNotifications': {
+        'key': 'workflow:notifications',
+        'default': UserOptionValue.all_conversations,  # '0'
+        'type': int,
+    }
+}
```
Comment (MaxBittker) [inline src/sentry/api/endpoints/user_notification_details.py]: it's slightly sad we have these constants defined twice :x we don't have a great solution today for sharing constants between python and JS though, but we could implement one in the build step.

File: src/sentry/api/endpoints/user_notification_details.py [hunk 2]
```diff
@@ -24,20 +56,44 @@ def get_attrs(self, item_list, user, *args, **kwargs):
...
+        raw_data = {option.key: option.value for option in attrs}
+
+        data = {}
+        for key in USER_OPTION_SETTINGS:
+            uo = USER_OPTION_SETTINGS[key]
+            val = raw_data.get(uo['key'], uo['default'])
+            if (uo['type'] == bool):
+                data[key] = bool(int(val))  # '1' is true, '0' is false
+            elif (uo['type'] == int):
+                data[key] = int(val)
```
Comment (MaxBittker) [inline src/sentry/api/endpoints/user_notification_details.py]: slightly worried me that things get silently dropped if their type isn't bool or int, but that paranoia might not be worth adding a raise or a log statement

File: src/sentry/api/endpoints/user_notification_details.py [hunk 3]
```diff
@@ -24,20 +56,44 @@ def get_attrs(self, item_list, user, *args, **kwargs):
...
+
+        return data
+
+
+class UserNotificationDetailsSerializer(serializers.Serializer):
+    deployNotifications = serializers.IntegerField(required=False, min_value=2, max_value=4)
```
Comment (MaxBittker) [inline src/sentry/api/endpoints/user_notification_details.py]: could be kind of nice/DRY to generate these directly from the option dictionary and put the ranges in there, but also could make the code grosser and not be worth it

File: src/sentry/api/endpoints/user_notification_details.py [hunk 4]
```diff
@@ -24,20 +56,44 @@ def get_attrs(self, item_list, user, *args, **kwargs):
...
         serialized = serialize(user, request.user, UserNotificationsSerializer())
         return Response(serialized)
+
+    def put(self, request, user):
+        serializer = UserNotificationDetailsSerializer(data=request.DATA)
+
+        if serializer.is_valid():
+            for key in serializer.object:
+                db_key = USER_OPTION_SETTINGS[key]['key']
+                val = six.text_type(int(serializer.object[key]))
```
Comment (MaxBittker) [inline src/sentry/api/endpoints/user_notification_details.py]: what is this type conversion for?

Comment (MaxBittker) [reply src/sentry/api/endpoints/user_notification_details.py]: does it go bool -> int -> string?

File: src/sentry/api/endpoints/user_notification_details.py [hunk 5]
```diff
@@ -24,20 +56,44 @@ def get_attrs(self, item_list, user, *args, **kwargs):
...
+                (uo, created) = UserOption.objects.get_or_create(user=user, key=db_key, project=None)
+                uo.update(value=val)
```
Comment (MaxBittker) [inline src/sentry/api/endpoints/user_notification_details.py]: not a good idea for this code, but i realized per our discussion earlier that you could achieve some degree of bulk updates by batching them by value.

**Review by MaxBittker — APPROVED** [review]
code and tests look good to me! left some thoughts and questions.

File: src/sentry/api/endpoints/user_notification_details.py (hunk 4 above)
Comment (lynnagara) [reply src/sentry/api/endpoints/user_notification_details.py]: Yeah, the fields that we're exposing as booleans here are actually being stored as strings '0' and '1'

**Pull request merged**